cel/decls.go

```diff
@@ -0,0 +1,1135 @@
+	return &Type{
+		kind:        OpaqueKind,
+		runtimeType: types.NewTypeValue(name),
+		parameters:  params,
+	}
+}
+
+// ObjectType creates a type references to an externally defined type, e.g. a protobuf message type.
+func ObjectType(typeName string) *Type {
+	return &Type{
+		kind:        StructKind,
+		runtimeType: types.NewObjectTypeValue(typeName),
+	}
+}
+
+// TypeParamType creates a parameterized type instance.
+func TypeParamType(paramName string) *Type {
+	return &Type{
+		kind:        TypeParamKind,
+		runtimeType: types.NewTypeValue(paramName),
+	}
+}
+
+// Variable creates an instance of a variable declaration with a variable name and type.
+func Variable(name string, t *Type) EnvOption {
+	return func(e *Env) (*Env, error) {
+		et, err := TypeToExprType(t)
+		if err != nil {
+			return nil, err
+		}
+		e.declarations = append(e.declarations, decls.NewVar(name, et))
+		return e, nil
+	}
+}
+
+// Function defines a function and overloads with optional singleton or per-overload implementations.
+//
+// Using Function is roughly equivalent to calling Declarations() to declare the function signatures
+// and Functions() to define the function implementations, if they have been defined. Specifying the
+// same function name more than once will result in the aggregation of the function overloads. If any
+// signatures conflict between the existing and new function definition an error will be raised.
+// However, if the signatures are identical and the overload ids are the same, the redefinition will
+// be considered a no-op.
+//
+// One key difference with using Function() is that each FunctionDecl provided will handle dynamic
+// dispatch based on the type-signatures of the overloads provided which means overload resolution at
+// runtime is handled out of the box rather than via a custom implementation for overload resolution
+// via Functions().
+func Function(name string, opts ...FunctionOpt) EnvOption {
+	return func(e *Env) (*Env, error) {
+		fn := &functionDecl{
+			name:      name,
+			overloads: map[string]*overloadDecl{},
+			options:   opts,
+		}
+		err := fn.init()
+		if err != nil {
+			return nil, err
+		}
+		_, err = functionDeclToExprDecl(fn)
+		if err != nil {
+			return nil, err
+		}
+		if existing, found := e.functions[fn.name]; found {
+			fn, err = existing.merge(fn)
+			if err != nil {
+				return nil, err
+			}
+		}
+		e.functions[name] = fn
+		return e, nil
+	}
+}
+
+// FunctionOpt defines a functional  option for configuring a function declaration.
+type FunctionOpt func(*functionDecl) (*functionDecl, error)
+
+// SingletonUnaryImpl creates a singleton function defintion to be used for all function overloads.
+//
+// Note, this approach works well if operand is expected to have a specific trait which it implements,
+// e.g. traits.ContainerType. Otherwise, prefer per-overload implementations.
+func SingletonUnaryImpl(fn functions.UnaryOp, traits ...int) FunctionOpt {
```

Comment:
Or is this to avoid collision?

Reply:
It's mostly just to avoid collision between `Function` as declaration and `Impl` as the binding. I could change it to `SingletonUnaryBinding` and so on if that sounds better.

Reply:
`Binding` does sound less "internal details" then `Impl`, but I'll leave it up to you.

Reply:
Done.